### PR TITLE
Allow custom attribute filtering with an open generic type.

### LINKED ISSUE #64169: [API Proposal]: Get info about generic attributes
### Background and motivation

Make it easier to handle generic attributes.

### API Proposal

- Allow use open generic as a param to `Attribute.GetCustomAttributes` Method.. It currently returns an empty array.
- ~~Add an the  instance property `TypeArgs` to the Attribute class  that returns all type args if any.~~

### API Usage

```C#
var Attrs = Attribute.GetCustomAttributes(this.GetType(), typeof(TestAttribute<>));
var type = Attrs[0].TypeArgs[0];
```

### Alternative Designs

_No response_

### Risks

_No response_

### PR DESCRIPTION
**Pull request by madelson**
Filtering on an open generic type retrieves all attributes that are or
are derived from constructed instances of that type.

fix #64169

### CODE REVIEW

**Comment by madelson** [comment]
@buyaa-n do any of the failures above look "real" to you?

**Comment by buyaa-n** [comment]
> @buyaa-n do any of the failures above look "real" to you?

Not sure what you mean by "real", but the failures looks unrelated.

Anyway assume your PR is not ready as it is a draft

**Comment by madelson** [comment]
@buyaa-n thanks. I was just looking for confirmation that I shouldn't be diving deeper into the check failures. I've removed the draft status.

File: src/coreclr/nativeaot/System.Private.CoreLib/src/Internal/Reflection/Extensions/NonPortable/CustomAttributeSearcher.cs [hunk 1]
```diff
@@ -74,6 +74,30 @@ public IEnumerable<CustomAttributeData> GetMatchingCustomAttributes(E element, T
                         return true;
                     };
             }
+            else if (optionalAttributeTypeFilter.IsGenericTypeDefinition)
```
Comment (madelson) [inline src/coreclr/nativeaot/System.Private.CoreLib/src/Internal/Reflection/Extensions/NonPortable/CustomAttributeSearcher.cs:77]: Question: elsewhere in this file there are comments about some reflection metadata (e. g. IsSealed) not being present. Might that be true for the generics-related properties as well? If so, how should we handle that case?

File: src/libraries/System.Runtime/tests/System/Attributes.cs [hunk 2]
```diff
@@ -295,10 +295,21 @@ public static void GetCustomAttributesWorksWithOpenAndClosedGenericTypesForEvent
...
+        [Fact]
+        [ActiveIssue("https://github.com/dotnet/runtime/issues/56887", TestRuntimes.Mono)]
+        public static void GetCustomAttributesOnOpenGenericTypeRetrievesDerivedAttributes()
+        {
+            Attribute[] attributes = Attribute.GetCustomAttributes(typeof(HasGenericAttribute), typeof(GenericAttribute<>));
+            Assert.Equal(2, attributes.Length);
+            Assert.Equal(1, attributes.OfType<DerivesFromGenericAttribute>().Count());
+            Assert.Equal(1, attributes.OfType<GenericAttribute<string>>().Count());
+        }
+
```
Comment (buyaa-n) [inline src/libraries/System.Runtime/tests/System/Attributes.cs:307]: NIT: Could we have a few more tests like for other member types also some scenarios like having same attribute with different type parameters and requesting an attribute with a concrete type would return only that requested one (of course if that scenario is not covered)

File: src/coreclr/nativeaot/System.Private.CoreLib/src/Internal/Reflection/Extensions/NonPortable/CustomAttributeSearcher.cs (hunk 1 above)
Comment (madelson) [reply src/coreclr/nativeaot/System.Private.CoreLib/src/Internal/Reflection/Extensions/NonPortable/CustomAttributeSearcher.cs:77]: @buyaa-n I'm not sure I follow.

`optionalAttributeTypeFilter.IsSealed` is accessed inside a try-catch and is "optional" in the sense that it is just used as an optimization so we get the exception it's not a problem.

In contrast, if we sometimes can't check whether we're a generic type definition, then what do we do if the user passes an open generic type?

Comment (buyaa-n) [reply src/coreclr/nativeaot/System.Private.CoreLib/src/Internal/Reflection/Extensions/NonPortable/CustomAttributeSearcher.cs:77]: Sorry, I misunderstood your initial comment 😝, IDK if the same issue could happen for `optionalAttributeTypeFilter.IsGenericTypeDefinition` @jkotas might help with that

**Review by buyaa-n — APPROVED** [review]
Except improving the test coverage, changes LGTM, thanks!

File: src/coreclr/nativeaot/System.Private.CoreLib/src/Internal/Reflection/Extensions/NonPortable/CustomAttributeSearcher.cs (hunk 1 above)
Comment (jkotas) [reply src/coreclr/nativeaot/System.Private.CoreLib/src/Internal/Reflection/Extensions/NonPortable/CustomAttributeSearcher.cs:77]: `IsGenericTypeDefinition` and other similar bool properties on System.Type won't throw `MissingMetadataException`.

**Review by jkotas — APPROVED** [review]

File: src/libraries/System.Runtime/tests/System/Attributes.cs (hunk 2 above)
Comment (madelson) [reply src/libraries/System.Runtime/tests/System/Attributes.cs:313]: > Could we have a few more tests like for other member types

We have tests for all the member types calling into `GenericAttributesTestHelper` which tests open generic attributes.

> same attribute with different type parameters and requesting an attribute with a concrete type would return only that requested one

Added more explicit coverage for this case.

Comment (buyaa-n) [reply src/libraries/System.Runtime/tests/System/Attributes.cs:313]: Great, thanks!

**Comment by buyaa-n** [comment]
Thanks @madelson for your contribution!

The failure unrelated, filed [an issue](https://github.com/dotnet/runtime/issues/68359), merging

**Pull request merged**